### PR TITLE
services, xds, orca: support EPS in client-side WRR

### PR DESCRIPTION
**Pull request by danielzhaotongliu**
Design: go/wrr-eps-design-review-slides

Uses: #10160

Tested: updated & ran tests 100 times to ensure they are not flaky

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 1]
```diff
@@ -184,8 +184,14 @@ final class WrrSubchannel extends ForwardingSubchannel {
...
     void onLoadReport(MetricReport report) {
-      double newWeight = report.getCpuUtilization() == 0 ? 0 :
-              report.getQps() / report.getCpuUtilization();
+      double newWeight = 0;
+      if (report.getCpuUtilization() > 0 && report.getQps() > 0) {
+        double penalty = 0;
+        if (report.getEps() > 0 && config.errorUtilizationPenalty > 0) {
+          penalty = report.getEps() / report.getQps() * config.errorUtilizationPenalty;
+        }
+        newWeight = report.getQps() / (report.getCpuUtilization() + penalty);
```
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:193]: if the cpu utilization is 0 and penalty is not, then the `newWeight` becomes 0, is that what we want?

Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:193]: From my understanding, yes. Since originally with the old formula `weight = qps / cpu_utilization`, when `cpu_utilization` is 0, weight is 0. Would the case of `cpu_utilization` is 0 and penalty is not 0 ever happen?

Perhaps @yousukseung can clarify and confirm.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 2]
```diff
@@ -184,8 +184,14 @@ final class WrrSubchannel extends ForwardingSubchannel {
 
     @VisibleForTesting
     void onLoadReport(MetricReport report) {
-      double newWeight = report.getCpuUtilization() == 0 ? 0 :
-              report.getQps() / report.getCpuUtilization();
+      double newWeight = 0;
+      if (report.getCpuUtilization() > 0 && report.getQps() > 0) {
+        double penalty = 0;
+        if (report.getEps() > 0 && config.errorUtilizationPenalty > 0) {
```
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: The config object is not thread safety. It is not volatile anymore, but actually the config update and this onLoadReport referring to it might be from different threads.

**Comment by YifeiZhuang** [comment]
> @YifeiZhuang PTAL. Thanks. Rebased

Were you able to request a reviewer if you are not a maintainer?
Otherwise I don't get notified.

The PR is mostly fine to me. Only a few comments.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 2 above)
Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: From my understanding, the entire `config` itself is only updated in `acceptResolvedAddresses()` and the fields (e.g. `config.errorUtilizationPenalty`) are final and cannot be modified/updated/set for all onLoadReport calls until another `acceptResolvedAddresses()` call.

Furthermore, `acceptResolvedAddresses()` then calls `afterAcceptAddresses()` which "re-register" the `onLoadReport` listeners. So the config field should not change for the lifetime of all `onLoadReport` calls?

**Comment by danielzhaotongliu** [comment]
> Were you able to request a reviewer if you are not a maintainer?
> Otherwise I don't get notified.

No. I am not able to request any reviewer in any opened PR (that option is greyed-out for me) since I am not a maintainer

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 2 above)
Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: The config update is in `syncContext`. The perPRC `onLoadReport` listener  is set to stream tracer when pick subchannel, also in `syncContext`.  The config change seen from the other thread (the transport thread) has thread safety issue - this can be stale data or random data.

Note that the `afterAcceptAddress()` is only for oob listener, which is different from perPpcListener. OOB listener is fine, because it is called back from a `syncContext`.

cc. @ejona86

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 1 above)
Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:193]: It looks to me an edge case. The difference is that in the rare case (cpu_utilization is 0 and penalty is not 0), the channel gets an average weight or (qps/penalty) weight. Document the result if you decide to ignore that rare case and treat it as cpu utilization = 0?

Comment (yousukseung) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:193]: We should use the average weight when `cpu_utilization` is zero as indicated in the gRFC. Also confirmed C++ does that.

gRFC:
```
function UpdateWeight(qps, eps, cpu_utilization) {
  if (cpu_utilization > 0 && qps > 0)
    cpu_utilization += eps / qps * config.error_utilization_penalty;
  var new_weight =
     cpu_utilization == 0 ? 0 : qps / cpu_utilization;
  if (new_weight == 0) return;
  if (non_empty_since_ == infy) non_empty_since_ = now();
  last_updated_ = now();
  weight_ = new_weight;
}
```


C++:
```
float weight = 0;
if (qps > 0 && cpu_utilization > 0) {
  double penalty = 0.0;
  if (eps > 0 && error_utilization_penalty > 0) {
    penalty = eps / qps * error_utilization_penalty;
  }
  weight = qps / (cpu_utilization + penalty);
}
```

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 2 above)
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: Pick subchannel is not in the `syncContext`, and the tracer isn't called within the `syncContext`. The problem is not the contents of config, but reading the `config` field itself.

Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: That makes sense. I see that `config` is not `volatile` as of #10139, what would be the reasoning for not making it `volatile`?

Could a fix be (I think approach 1 would be better):

1.  add a separate `volatile` variable/field (in this class) for error utilization penalty and update this variable whenever the `config` is updated (after line 100-101). Although this does bring a slight cost of maintaining a `volatile` variable. 

2. use `synchronized` on the method but this incurs a heavy performance penalty with locking and since we are only reading the fields and `onLoadReport` could be called frequently, so **I don't think this would be a good approach**.

On a related note, I see that there are other reads of the `config` field in this file (e.g. in the `getWeight()` method), wouldn't those be currently affected as well since the `config` object itself is not `volatile`?

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 1 above)
Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:193]: Thanks for the clarification. The current Java code follows the same behaviour as C++ and the [gRFC A58](https://github.com/grpc/proposal/blob/master/A58-client-side-weighted-round-robin-lb-policy.md).

```
double newWeight = 0;
if (report.getCpuUtilization() > 0 && report.getQps() > 0) {
  double penalty = 0;
  if (report.getEps() > 0 && config.errorUtilizationPenalty > 0) {
    penalty = report.getEps() / report.getQps() * config.errorUtilizationPenalty;
  }
  newWeight = report.getQps() / (report.getCpuUtilization() + penalty);
}
if (newWeight == 0) {
  return;
}
if (nonEmptySince == infTime) {
  nonEmptySince = ticker.nanoTime();
}
lastUpdated = ticker.nanoTime();
weight = newWeight;
```

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 2 above)
Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: Yes I'll correct that pick subchannel is not always in `syncContext` (it looks pending stream reprocess picking subchannel is in `syncContext`, but not directly pick when there is already a picker)

I agree that adding to the `errorUtilizationPenalty ` config to the  constructor is feasible. We already did that with other config (`config.enableOobLoadReport`)

Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:190]: Other places like `getWeight()` is called within `syncContext` - the timer is scheduled from there so it is fine.

`per-rpc-listener` for the load reports is final per each subchannel, the stream tracer is constructed with the listener, so it looks fine for its thread safety.

Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java]: Done.
* Added `config.errorUtilizationPenalty` to the picker, then passed to the sub-channel, allocate a new listener each RPC (only for `perRpcListener` since OOB listener is already thread safe)
* Updated tests & improved coverage for edge cases not tested previously

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 3]
```diff
@@ -172,8 +173,8 @@ public Subchannel createSubchannel(CreateSubchannelArgs args) {
   @VisibleForTesting
   final class WrrSubchannel extends ForwardingSubchannel {
     private final Subchannel delegate;
-    private final OrcaOobReportListener oobListener = this::onLoadReport;
-    private final OrcaPerRequestReportListener perRpcListener = this::onLoadReport;
+    private final OrcaOobReportListener oobListener = new OrcaReportListener(
```
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:176]: Here the oobListener should not be constructed like this - because it won't get the updated config for the whole lifetime of the subchannel. (as we talked before, each address update in the LB with `acceptResolvedAddresses`, some subchannels can be reused if their address is the same). 
For the optimized oobListener (avoid creating frequently), it should use the `config` object directly  in `onLoadReport`(so that it uses the updated config and it is also thread safe because `onLoadReport` is called within `syncContext`), but not constructed with the current `config.errorUtilizationPenalty` value.

Since the oobListener is set per each address update worse case, not as frequent as pick subchannel, you can just create a new ooblistener each time needed.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 4]
```diff
@@ -251,7 +269,7 @@ public PickResult pickSubchannel(PickSubchannelArgs args) {
       if (!enableOobLoadReport) {
         return PickResult.withSubchannel(subchannel,
             OrcaPerRequestUtil.getInstance().newOrcaClientStreamTracerFactory(
-                ((WrrSubchannel)subchannel).perRpcListener));
+                ((WrrSubchannel)subchannel).new OrcaReportListener(errorUtilizationPenalty)));
```
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:272]: Here the per-request listener can be created once at picker's construction time and reused.
A followup PR is fine.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 3 above)
Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java]: Done.
Created a new OOB listener each time needed.

Thanks for the detailed context and explanation, I previously thought the OOB listener is completely separated from the lifecycle of channel & picker.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 4 above)
Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java]: Done.
The per-request listener is created once (per channel & tracked with a map) in the picker's constructor.
I added a "fallback" case with `getOrDefault` for safety.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 2 above)
Comment (danielzhaotongliu) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java]: Please see updated commits & comment thread below.

**Review by YifeiZhuang — APPROVED** [review]
LGTM.
